### ISSUE TITLE
Split gene identification in importer in 3 runs: first by Entrez, then by symbol, then by alias

### ISSUE THREAD

**Issue by sandertan** [cBioPortal/cbioportal#4615]
Recently @angelicaochoa pushed an update to the importer to retrieve the database gene by Symbol if the Entrez ID in data is not found: https://github.com/cBioPortal/cbioportal/pull/4114 . I think this is a welcome change, because it allows more data to be loaded into the portal. However, this could lead to more incorrect gene mappings.

First of all, my concern assumes that the community has previously decided that Entrez ID is the leading gene ID when importing data in cBioPortal. Personally, I think the gene symbol, HGNC currated (www.genenames.org), should be the leading gene ID, since this is what the user will query in the user interface. Entrez ID is nowhere to be found in the front end.

#### Example of where gene mapping goes wrong:
Hypothetical `gene` table in database:

| Entrez ID | Symbol |
|-----------|--------|
| 10        | GENE_X |
| 11        | GENE_Y |

Hypothetical `gene_alias` table in database:

| Entrez ID | Alias  |
|-----------|--------|
| 10        | GENE_Y |

Hypothetical gene expression data file:

| Entrez ID | Symbol | SAMPLE_1 |
|-----------|--------|----------|
| 1         | GENE_Y | 0        |
| 10        | GENE_X | 15       |

When this data is imported, the first entry will be parsed, and `Entrez ID 1` is not found in the database so symbol `GENE_Y` is used. Through the alias table, this maps the entry to `Entrez ID 10`.

When the second entry is parsed, GENE_X can be loaded into the database, but this has the same Entrez ID as the first entry. I'm not sure if this 2nd entry is skipped or that it overwrites the 1st entry. But regardless of what happens, if Entrez ID should be leading and the first entry should have been skipped.

#### Existing problem
This problem was already present with gene aliases and symbols. If the Entrez ID column was missing, genes were imported based on Symbol. If the symbol was not found in the `gene` table, the symbols in the `gene_alias` were used. If at a later point in the data, the official symbol was found, it maps to the same Entrez ID.

#### Solution 
Regardless of the order of rows, if the Entrez ID is leading, all data should be imported based on Entrez ID first. For all data entries that are not mapped to a database Entrez ID, a second import process should be started in an attempt to match these to symbols in the `gene` table. For all data entries that are still not mapped to a gene, a third import process should start and use the symbols in the `gene_alias` table.

**Comment by jjgao** [comment]
@sandertan thanks for creating this ticket. I agree the scenario can theoretically happen, but hopefully, it's a rare situation. Do you have any example in the datahub?

I think the better option is detecting the wrong entrez ID and correct them in the file.

**Comment by sandertan** [comment]
@jjgao Yes this happens quite often. For example, for data_RNA_Seq_v2_expression_median.txt in brca_tcga:

Total datalines: 20541
Old skipped lines: 118
New skipped lines: 87

In the new importer out of the 87 skipped lines:
- Skipped lines due to record not found: 51
- Skipped lines due to gene duplication: 34 (<- this is the case I'm describing)
- Skipped lines due to Symbol ambiguity: 2

#### Importer reports
[report_brca_tcga_old.txt](https://github.com/cBioPortal/cbioportal/files/2280012/report_brca_tcga_old.txt)
[report_brca_tcga_new.txt](https://github.com/cBioPortal/cbioportal/files/2280011/report_brca_tcga_new.txt)
[brca_tcga_gitdiff.txt](https://github.com/cBioPortal/cbioportal/files/2280010/brca_tcga_gitdiff.txt)

#### Example
AGAP4 & AGAP8 is one of the cases where this happens, and which I described in #2498
```
Gene AGAP4 (119016)(given as alias in your file as: AGAP8)  found to be duplicated in your file. Duplicated row will be ignored!; 1x
```

**Comment by fedde-s** [comment]
When the validation script was first prepended into the data loading pipeline, the plan was to allow data curators to make the data file unambiguous before then, during the ‘transform’ step of the ET(v)L.

Would it be feasible to facilitate this somehow? To use information from the database to format an a data file that yields the exact same result as the procedure you describe, by including unambiguous Entrez IDs? It is for this use-case that the validator warns someone who wants to load data to check whether they intended the file to be ambiguous.

One, probably insufficient, option intended to provide transformation code with such information, is the ‘error file’ option of the validation script.

**Comment by fedde-s** [comment]
One of the design goals we had during the ET(v)L separation was for the data loading code itself to only concern itself with the actual loading, and contain very little logic that makes assumptions about what the data curators may have forgotten to do during the transformation step.

**Comment by pieterlukasse** [comment]
Thanks @sandertan, @jjgao. Here are some of my thoughts on this. The change proposed in #4114 to prioritize hugo symbol over entrez id (albeit only in some cases) can be considered a paradigm change for the loader until now. In my opinion, this change makes the entrez id value stored in the DB unreliable, since one cannot trust anymore that this was really the entrez id found in the original data. I would argue that given such unreliability, the entrez gene id becomes useless. Therefore, the question I would like to ask is: can we live with a design where we ignore entrez gene id completely? If yes, what would be the consequences and drawbacks (if any)? If no, why should we allow it for a subset of the data? 

To better answer these questions, I think we need to find out why the entrez gene ids are changing in the first place. Take for example the recent case `MICA` (from https://www.ncbi.nlm.nih.gov/gene/4276 to https://www.ncbi.nlm.nih.gov/gene/100507436). Why was the entrez gene id deprecated? We need to find out first with NCBI first what are *all* the possible causes of deprecation and what are the *most* common ones. Until then, I think we should put any change on hold.
